# The Python Mega Course: Build 10 Real World Applications
---

This notebook contains the source code for the video lectures of Section 3 of [The Python Mega Course: Build 10 Real World Applciations](https://www.udemy.com/the-python-mega-course/?couponCode=GITHEADSECTION).

# Section 3: Beyond the Basics
***

**Lecture 59:** [Section Introduction](https://www.udemy.com/the-python-mega-course/learn/v4/t/lecture/9542698?start=0)
---

This section covers some more advanced concepts of Python such as while loops, date and time objects, advanced file namipulations, and more.

**Lecture 60:** [Setting Up](https://www.udemy.com/the-python-mega-course/learn/v4/t/lecture/9542704?start=180)
---

This lecture covers how to setup Python and the Atom editor on Windows, Mac, and Linux.

**Lecture 61:** [While Loop](https://www.udemy.com/the-python-mega-course/learn/v4/t/lecture/9529658?start=0)
---

In [58]:
#The loop below will run as long as x is less than 3
x = 0
while x < 3:
    print("Smaller")
    x = x + 1

Smaller
Smaller
Smaller


**Lecture 62:** [While Loop with Password Checker](https://www.udemy.com/the-python-mega-course/learn/v4/t/lecture/9542712?start=0)
---

In [59]:
correct_password = "python123"
password = input("Enter password: ")

#the loop below will run as long as a wrong password is submitted
while correct_password != password:
    password = input("Wrong password! Enter again: ")
    
print("Logged in")

Enter password: python123
Logged in


**Lecture 63:** [String Formatting](https://www.udemy.com/the-python-mega-course/learn/v4/t/lecture/9529762?start=0)
---

In [1]:
correct_password = "python123"
name = input("Enter name: ")
surname = input("Enter surname: ")
password = input("Enter password: ")

while correct_password  != password:
    password = input("Wrong password! Enter again: ")

#the %s will be replaced by the values of name and surname
message = "Hi %s %s, you are logged in" % (name, surname) 
print(message)


Enter name: Ardit
Enter surname: Sulce
Enter password: python123
Hi Ardit Sulce, you are logged in


**Lecture 64:** [Modules, Libraries, and Packages](https://www.udemy.com/the-python-mega-course/learn/v4/t/lecture/9529672?start=0)
---

In [2]:
#Example of importing a module
import os

In [6]:
#Example of importing a package
import sqlite3

In [7]:
#get module installation path
os.__file__

'/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/os.py'

In [9]:
#get package installation path
sqlite3.__file__

'/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/sqlite3/__init__.py'

Modules are made of just one .py file. Packages are made of multiple files arranged in one or more folders. If \__file\__ returns the name of the imported object then it is a module. If it returns \__init\__.py it is a package. A library is usually a word used to refer to either a module or a package.

**Lecture 65:** [Installing Python Packages](https://www.udemy.com/the-python-mega-course/learn/v4/t/lecture/9542632?start=0)
---

To install third-party libraries use pip in your command line. For example, to install the library *glob2* you do:

In [ ]:
pip install glob2

If you are using Python3 on Mac or Linux, you may want to do this instead:


In [ ]:
pip3 install glob2

If none of the above doesn't work you may try one of the following:

In [ ]:
python -m pip install glob2

In [ ]:
python3 -m pip install glob2

**Lecture 66:** [Working with Date and Time Objects](https://www.udemy.com/the-python-mega-course/learn/v4/t/lecture/9529654?start=0)
---

In [16]:
#import the datetime class from the datetime module
from datetime import datetime

In [17]:
#Calculate the difference between now and date 12/31/1900
delta = datetime.now() - datetime(1900, 12, 31)
delta

datetime.timedelta(42901, 59738, 86244)

In [18]:
#Difference in days
delta.days

42901

In [19]:
#The remainder in seconds
delta.seconds

59738

So, the difference between now and 12/31/1900 is 42901 days and 59442 seconds

In [24]:
#create a datetime object manually
then = datetime(1900, 12, 31, 20, 12, 59, 83845)
then

datetime.datetime(1900, 12, 31, 20, 12, 59, 83845)

The above means *datetime(year, month, day, hour, minute, second, microsecond)*

In [25]:
#create a datetime object representing the now
now = datetime.now()
now

datetime.datetime(2018, 6, 16, 16, 41, 35, 438543)

In [29]:
#create a datetime object out of a string
whenever = datetime.strptime("2017-12-31", "%Y-%m-%d")
whenever

datetime.datetime(2017, 12, 31, 0, 0)

In [30]:
#create a datetime object out of string, including hour, minutes, and seconds
whenever = datetime.strptime("2017:12:31:20:59", "%Y:%m:%d:%H:%M")
whenever

datetime.datetime(2017, 12, 31, 20, 59)

In [31]:
#create a string out of a datetime object
whenever.strftime("%Y-%m-%d %H:%M")

'2017-12-31 20:59'

[Here](http://strftime.org) is a complete list of srftime symbols. 

Use *dir(whenever)* to get a list of methods and properties applied to the datetime object. Examples are *whenever.year*, *whenever.month*, *whenever.day*, etc.

**Lecture 67:** [Complete List of Datetime Formatting Codes](https://www.udemy.com/the-python-mega-course/learn/v4/t/lecture/9550362?start=0)
---

Again, this lecture contains the complete list of srtfime formatting codes.

**Lecture 68:** [Iterating Multiple Sequences](https://www.udemy.com/the-python-mega-course/learn/v4/t/lecture/9529662?start=0)
---

In [35]:
#accessing items from two sequences at the same time
a = ['a', 'b', 'c']
b = [1, 2, 3]
for i, j in zip(a, b):
    print(i, j)

a 1
b 2
c 3


In [37]:
# a more advanced version
for i, j in zip(a, b):
    print("%s is %s" % (i, j))

a is 1
b is 2
c is 3


**Lecture 69:** [The "with" Context Manager](https://www.udemy.com/the-python-mega-course/learn/v4/t/lecture/9529652?start=0)
---

In [39]:
#opening a file using the standard way
myfile = open("example.txt", "w")
myfile.write("Something")
myfile.close()

In [40]:
#opening a file using the "with" context manager
with open("example.txt", "w") as myfile:
    myfile.write("Something")

**Lecture 70:** [Exercise](https://www.udemy.com/the-python-mega-course/learn/v4/t/lecture/5173680?start=0)
---

Please take a look at the following code:

In [ ]:
temperatures = [10, -20, -289, 100]
def c_to_f(c):
    if c < -273.15:
        return "That temperature doesn't make sense!"
    else:
        f = c* 9/5 + 32
        return f
for t in temperatures:
    print(c_to_f(t))

The code prints out the output of the *c_to_f*  function multiple times in the terminal. 

For this exercise, please write the output in a text file instead of printing it out in the terminal.

The text file content will look like this:<br>

50.0<br>
-4.0<br>
212.0<br>

Please don't write any message in the text file when input is lower than -273.15.

**Lecture 71:** [Solution](https://www.udemy.com/the-python-mega-course/learn/v4/t/lecture/5173682?start=0)
---

In [42]:
#one solution
temperatures = [10,-20,-289,100]
 
def writer(temperatures):
    with open("temps.txt", 'w') as file:
        for c in temperatures:
            if c > -273.15:
                f = c* 9/5 + 32
                file.write(str(f) + "\n")
 
writer(temperatures) #Here we're calling the function, otherwise no output will be generated

In [43]:
#an even better solution
temperatures = [10,-20,-289,100]
 
def writer(temperatures, filepath):
    with open(filepath, 'w') as file:
        for c in temperatures:
            if c > -273.15:
                f = c* 9/5 + 32
                file.write(str(f) + "\n")
 
writer(temperatures, "temps.txt") #Here we're calling the function, otherwise no output will be generated

**Lecture 72 -73:** [Exercise](https://www.udemy.com/the-python-mega-course/learn/v4/t/lecture/5220516?start=0)
---

Here is a tricky exercise.

1. Please download [these three files](http://pythonhow.com/data/3files/) and put them in a folder. The first text file contains the text *Content1* . The second contains *Content2* and the third *Content3* .
<br>
2. You should create a Python script that generates a new text file which should contain the content of all three text files. So the generated file should have this content:
<br>
*Content1 <br>
Content2 <br>
Content3* <br>

In other words, your Python script should merge the three text files. 
<br>
3. Also, the name of the output file should be the current timestamp. Example: *2017-11-01-13-57-39-170965.txt* 

You have some tips in the next lecture and the solution in the lecture after that.

**Lecture 74:** [Solution](https://www.udemy.com/the-python-mega-course/learn/v4/t/lecture/5220572?start=0)
---

In [45]:
import glob2
from datetime import datetime
 
filenames = glob2.glob("*.txt")
with open(datetime.now().strftime("%Y-%m-%d-%H-%M-%S-%f")+".txt", 'w') as file:
    for filename in filenames:       
        with open(filename, "r") as f:
            file.write(f.read() + "\n")